desktop-src/WES/defining-events.md

```diff
@@ -12,99 +12,91 @@ Providers must define all the events that they write. To define an event, use th
 
 The **value** attribute is the event identifier and it must be unique for the events that you define. Whether you set the other attributes depends on who will be consuming the events and from where. If administrators will be consuming your events using a tool like Windows Event Viewer, then you must set the **channel** attribute. If the channel type is Admin, then you must also specify the **level** attribute and set it to one of the levels defined in Winmeta.xml (win:Critical through win:Informational).
 
-If the event contains event-specific data, you must set the **template** attribute to the identifier of the template that defines the event-specific data. The **level**, **keywords**, **task**, and **opcode** attributes are used to group or bucket events. Although these attributes are optional, you should consider specifying level, task, and opcode, so that consumers can easily access only those events of interest. The **level** and **keywords** attributes can also be used by an ETW tracing session to limit the events that are written to the event tracing log file. The **keywords** attribute contains a space-delimited list of keyword names defined in the manifest.
+If the event contains event-specific data, you must set the **template** attribute to the identifier of the template that defines the event-specific data. The **level**, **keywords**, **task**, and **opcode** attributes are used to group or bucket events. Although these attributes are optional, you should consider specifying level, task, opcode, and keywords, so that consumers can easily access only those events of interest. The **level** and **keywords** attributes can also be used by an ETW tracing session to limit the events that are written to the event tracing log file. The **keywords** attribute contains a space-delimited list of keyword names defined in the manifest. If multiple keywords are specified their mask values are OR'ed together to create the keyword value that the event will use.
```

Comment:
I would be stronger in recommending level and keyword.

Indicate that if no level is set, the default of 0 means that the event cannot be filtered based on level. Give examples of levels: win:Critical, win:Error, win:Warning, win:Informational, win:Verbose, and indicate that events should default to win:Verbose. Indicate that if no keyword is set, the default of 0 means that the event cannot be filtered based on level and will be rejected by providers that set the EVENT_ENABLE_PROPERTY_IGNORE_KEYWORD_0 property. Also, should be "keywords defined in the provider", not "keyword names defined in the manifest", and perhaps should have a link to the <keyword> element.

Maybe split this into multiple paragraphs - one introducing the optional attributes, one describing level and keyword (or one for level and one for keyword), and one describing task and opcode.

Should we mention the other attributes just so people know they exist (so they can investigate further if needed)? "name" (new), "attributes" (new), "version", "notLogged".

Reply:
FWIW, defining custom opcodes is rare for Crimson, but can be done to apply a logical "event kind" to the event. Perhaps something like "Function Entry" and "Function Exit" would be a reasonable example of a custom opcode.

Defining custom tasks happens more frequently. Usually people use it to associate a localized string to an event. Basically tasks seem to be mostly used for the purpose of assigning a decoder-visible localized name to an event (since we didn't support real event names until recently).

Reply:
This is specifically Event Log documentation. I don't want to get into ETW features like level or keywords or keyword 0 here, because the event log does not expose those things to the user.

I'd strongly advocate for moving this whole set of docs to the ETW portion of the documentation and then revising them to detail more about ETW content such as sessions and keyword/level filtering, however, I'm not going to do that in this change.